### ISSUE TITLE
New rule suggestion: Disallow overloads that all return the same type

### ISSUE THREAD

**Issue by tom-sherman** [typescript-eslint/typescript-eslint#2270]
From the [Typescript docs](https://www.typescriptlang.org/docs/handbook/functions.html#overloads) (emphasis mine):

> It’s not uncommon for a single JavaScript function to **return different types** of objects based on the shape of the arguments passed in.

Something I pick up on in code reviews frequently is using overloads simply to achieve different argument shapes without affecting the return type. I'd love a rule that would disallow overloads that all return the same type as I think there is very often a better solution (discriminated unions, different argument types, tuples).

## Example

This would be an error because all of the overload definitions return the same type.

```typescript
declare function pickCard(x: {suit: string; card: number; }[]): Card;
declare function pickCard(x: number): Card;
```

I think some examples (such as the above) could have an autofix as well but this may end up being too opinionated and I don't know if it's possible to have a rule autofix in some scenarios but not others.

**Comment by bradzacher** [comment]
Have you tried unified-signatures? I think this will do what you want.

https://www.github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin/docs/rules/unified-signatures.md

**Comment by tom-sherman** [comment]
Wow that is exactly what I'm looking for, great stuff!
